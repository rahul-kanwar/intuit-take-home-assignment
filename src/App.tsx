import React, { useState, useEffect } from 'react';
import CommentForm from './Components/CommentForm/CommentForm';
import CommentList from './Components/CommentList/CommentList';
import Notification from './Components/Notification/Notification';
import styled from 'styled-components';
import { Comment } from './api';
import io from 'socket.io-client';

const Container = styled.div`
  max-width: 600px;
  margin: auto;
  padding: 20px;
`;

const App: React.FC = () => {
  const [newComment, setNewComment] = useState<Comment | null>(null);
  const [notification, setNotification] = useState<string | null>(null);

  useEffect(() => {
    const socket = io('http://localhost:3001');

    socket.on('newComment', (comment: Comment) => {
      setNewComment(comment);
      setNotification('New comment added!');
    });

    return () => {
      socket.disconnect();
    };
  }, []);

  const handleCommentAdded = (comment: Comment) => {
    setNewComment(comment);
    setNotification('New comment added!');
  };

  const handleCloseNotification = () => {
    setNotification(null);
  };

  return (
    <Container>
      <h1 style={{textAlign: 'center'}}>Comments Feed</h1>
      <CommentForm onCommentAdded={handleCommentAdded} />
      {notification && <Notification message={notification} onClose={handleCloseNotification} />}
      <CommentList newComment={newComment} />
    </Container>
  );
};

export default App;
