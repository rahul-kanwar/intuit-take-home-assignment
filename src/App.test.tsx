import React from 'react';
import { render, waitFor, fireEvent, screen } from '@testing-library/react';
import 'jest-styled-components';
import { format } from 'date-fns';

import App from './App';
import {
  Comment,
  getComments as mockGetComments,
  createComment as mockPostComments,
} from './api';

// Mock comments data
const mockComments: Comment[] = [
  {
    id: 1,
    name: 'John Doe',
    message: 'This is a test comment',
    created: '2024-06-26T10:35:42.457Z',
  },
  {
    id: 2,
    name: 'Jane Doe',
    message: 'Another test comment',
    created: '2024-06-26T11:35:42.457Z',
  },
];

const mockNewComment: Comment = {
  id: 3,
  name: 'New User',
  message: 'This is a new comment',
  created: '2024-06-26T12:35:42.457Z',
};

// Mock the api functions
jest.mock('./api', () => ({
  getComments: jest.fn(),
  createComment: jest.fn(),
}));

describe('App Component', () => {
  beforeEach(() => {
    (mockGetComments as jest.Mock).mockResolvedValue(mockComments);
    (mockPostComments as jest.Mock).mockResolvedValue(mockNewComment);
  });
  it('renders correctly', async () => {
    const { asFragment } = render(<App />);

    await waitFor(() => {
      mockComments.forEach(comment => {
        expect(screen.getByText(comment.name)).toBeInTheDocument();
        expect(screen.getByText(comment.message)).toBeInTheDocument();
        expect(
          screen.getByText(format(new Date(comment.created), 'Pp'))
        ).toBeInTheDocument();
      });
    });

    expect(asFragment()).toMatchSnapshot();

    // Add a new comment
    fireEvent.change(screen.getByPlaceholderText('Name'), {
      target: { value: 'New User' },
    });
    fireEvent.change(screen.getByPlaceholderText('Comment'), {
      target: { value: 'This is a new comment' },
    });
    fireEvent.click(screen.getByText('Add Comment'));

    await waitFor(() => {
      expect(screen.getByText('New User')).toBeInTheDocument();
      // eslint-disable-next-line testing-library/no-wait-for-multiple-assertions
      expect(screen.getByText('This is a new comment')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
