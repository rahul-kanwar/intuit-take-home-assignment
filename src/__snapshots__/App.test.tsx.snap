// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders correctly 1`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  flex-direction: column;
  margin-bottom: 20px;
}

.c2 {
  margin-bottom: 10px;
  padding: 10px;
  font-size: 16px;
}

.c3 {
  margin-bottom: 10px;
  padding: 10px;
  font-size: 16px;
}

.c4 {
  padding: 10px;
  font-size: 16px;
  background-color: #007bff;
  color: white;
  border: none;
  cursor: pointer;
}

.c6 {
  border-bottom: 1px solid #ccc;
  padding: 10px 0;
}

.c5 {
  border-top: 1px solid #ccc;
  padding-top: 20px;
}

.c0 {
  max-width: 600px;
  margin: auto;
  padding: 20px;
}

<div
    class="c0"
  >
    <h1
      style="text-align: center;"
    >
      Comments Feed
    </h1>
    <form
      class="c1"
    >
      <input
        class="c2"
        placeholder="Name"
        required=""
        type="text"
        value=""
      />
      <textarea
        class="c3"
        placeholder="Comment"
        required=""
        rows="10"
      />
      <button
        class="c4"
        type="submit"
      >
        Add Comment
      </button>
    </form>
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        <h4>
          Jane Doe
        </h4>
        <p>
          Another test comment
        </p>
        <small>
          06/26/2024, 7:35 AM
        </small>
      </div>
      <div
        class="c6"
      >
        <h4>
          John Doe
        </h4>
        <p>
          This is a test comment
        </p>
        <small>
          06/26/2024, 6:35 AM
        </small>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App Component renders correctly 2`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  flex-direction: column;
  margin-bottom: 20px;
}

.c2 {
  margin-bottom: 10px;
  padding: 10px;
  font-size: 16px;
}

.c3 {
  margin-bottom: 10px;
  padding: 10px;
  font-size: 16px;
}

.c4 {
  padding: 10px;
  font-size: 16px;
  background-color: #007bff;
  color: white;
  border: none;
  cursor: pointer;
}

.c7 {
  border-bottom: 1px solid #ccc;
  padding: 10px 0;
}

.c6 {
  border-top: 1px solid #ccc;
  padding-top: 20px;
}

.c5 {
  position: fixed;
  top: 10px;
  opacity: 0.8;
  width: 320px;
  background-color: #4caf50;
  color: white;
  padding: 15px;
  border-radius: 5px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  z-index: 1000;
}

.c0 {
  max-width: 600px;
  margin: auto;
  padding: 20px;
}

<div
    class="c0"
  >
    <h1
      style="text-align: center;"
    >
      Comments Feed
    </h1>
    <form
      class="c1"
    >
      <input
        class="c2"
        placeholder="Name"
        required=""
        type="text"
        value=""
      />
      <textarea
        class="c3"
        placeholder="Comment"
        required=""
        rows="10"
      />
      <button
        class="c4"
        type="submit"
      >
        Add Comment
      </button>
    </form>
    <div
      class="c5"
    >
      New comment added!
    </div>
    <div
      class="c6"
    >
      <div
        class="c7"
      >
        <h4>
          New User
        </h4>
        <p>
          This is a new comment
        </p>
        <small>
          06/27/2024, 3:10 AM
        </small>
      </div>
      <div
        class="c7"
      >
        <h4>
          Jane Doe
        </h4>
        <p>
          Another test comment
        </p>
        <small>
          06/26/2024, 7:35 AM
        </small>
      </div>
      <div
        class="c7"
      >
        <h4>
          John Doe
        </h4>
        <p>
          This is a test comment
        </p>
        <small>
          06/26/2024, 6:35 AM
        </small>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
